Foundation/Network/IPv6.hs

```diff
@@ -0,0 +1,157 @@
+        )
+      where
+        ptr' :: Ptr (BE Word16)
+        ptr' = castPtr ptr
+    poke ptr ipv6 = do
+        let (i1,i2,i3,i4,i5,i6,i7,i8) = toTuple ipv6
+         in pokeOff ptr' 1 (toBE i1)
+         >> pokeOff ptr' 2 (toBE i2)
+         >> pokeOff ptr' 3 (toBE i3)
+         >> pokeOff ptr' 4 (toBE i4)
+         >> pokeOff ptr' 5 (toBE i5)
+         >> pokeOff ptr' 6 (toBE i6)
+         >> pokeOff ptr' 7 (toBE i7)
+         >> pokeOff ptr' 8 (toBE i8)
+      where
+        ptr' :: Ptr (BE Word16)
+        ptr' = castPtr ptr
+instance StorableFixed IPv6 where
+    size      _ = (size      (Proxy :: Proxy Word64)) `scale` 2
+    alignment _ = alignment (Proxy :: Proxy Word64)
+
+-- | serialise to human readable IPv6
+--
+-- >>> toString (fromString "0:0:0:0:0:0:0:1" :: IPv6)
+toString :: IPv6 -> String
+toString = fromList . toLString
+
+toLString :: IPv6 -> [Char]
+toLString ipv4 =
+    let (i1,i2,i3,i4,i5,i6,i7,i8) = toTuple ipv4
+     in intercalate ":" $ showHex4 <$> [i1,i2,i3,i4,i5,i6,i7,i8]
+
+showHex4 :: Word16 -> [Char]
+showHex4 c
+  | c < 0x10   = '0':'0':'0':showHex c
+  | c < 0x100  = '0':'0':showHex c
+  | c < 0x1000 = '0':showHex c
+  | otherwise  = showHex c
+
+showHex :: Word16 -> [Char]
+showHex = Base.printf "%03x"
+
+fromLString :: [Char] -> IPv6
+fromLString s =
+    case parseInt' (parseHex $ snd $ span isSeparator s) [] of
+        Left err        -> error err
+        Right (addr, _) -> addr
+    where
+        parseInt' :: Either [Char] (Word16, [Char])
+                  -> [Word16]
+                  -> Either [Char] (IPv6, [Char])
+        parseInt' (Left err)           _          = Left err
+        parseInt' (Right (w8, xs))     [w7,w6,w5,w4,w3,w2,w1] = Right (fromTuple (w1, w2, w3, w4, w5, w6, w7, w8), xs)
+        parseInt' (Right (_,  []))     _          = Left "Not an ipv6 addr"
+        parseInt' (Right (w,  ':':xs)) acc        = parseInt' (parseHex xs) (w:acc)
+        parseInt' (Right (_,    c:_ )) _          = Left $ "Not an ipv6 addr: unexpected char '" <> [c] <> "'"
+
+        parseHex :: [Char]
+                 -> Either [Char] (Word16, [Char])
+        parseHex buf =
+            case span isHexDigit buf of
+                ([], x:xs) -> case x of
+                                ':' -> Right (0, x:xs)
+                                _   -> Left $ "Not an ipv6 addr: unexpected char '" <> [x] <> "'"
+                (l , xs)   -> let lhs = readHex l :: [(Word16, [Char])]
+                              in  case lhs of
+                                    [(w, [])] -> Right (w, xs)
+                                    _ -> Left "can't fall here"
+
+
+-- | create an IPv6 from the given tuple
+fromTuple :: (Word16, Word16, Word16, Word16, Word16, Word16, Word16, Word16)
+          -> IPv6
+fromTuple (i1, i2, i3, i4, i5, i6, i7, i8) = IPv6 hi low
+  where
+    f :: Word16 -> Word64
+    f = fromIntegral
+    hi, low :: Word64
+    hi =    (f i1 .<<. 48) .&. 0xFFFF000000000000
```

Comment:
there's no need to mask after shifting, a Word16 is guarantee to be in the mask `0xFFFF`